pkg/lightning/backend/local/local.go

```diff
@@ -1942,142 +2486,107 @@ func (local *local) EngineFileSizes() (res []backend.EngineFileSize) {
+		return w.appendRowsSorted(kvs)
+	}
+	return w.appendRowsUnsorted(ctx, kvs)
 }
 
-func (w *Writer) writeRowsLoop() {
-	defer func() {
-		if w.writer != nil {
-			w.writer.cleanUp()
-			w.writer = nil
-		}
-		w.consumeCh <- struct{}{}
-	}()
-	var err error
-outside:
-	for {
-		w.flushChMutex.RLock()
-		flushCh := w.flushCh
-		w.flushChMutex.RUnlock()
-
-		select {
-		case kvs, ok := <-w.kvsChan:
-			if !ok {
-				break outside
-			}
-
-			w.writeBatch.append(kvs)
-			if w.writeBatch.totalSize <= w.memtableSizeLimit {
-				break
-			}
-			if w.writer == nil {
-				w.writer, err = newSSTWriter(w.genSSTPath())
-				if err != nil {
-					w.writeErr.Set(err)
-					return
-				}
-			}
-
-			if err = w.writeKVsOrIngest(0); err != nil {
-				w.writeErr.Set(err)
-				return
-			}
+func (w *Writer) flush(ctx context.Context) error {
+	w.Lock()
+	defer w.Unlock()
+	if w.batchCount == 0 {
+		return nil
+	}
 
-		case replyErrCh := <-flushCh:
-			err = w.writeKVsOrIngest(localIngestDescriptionFlushed)
-			if w.writer != nil {
-				err = w.writer.ingestInto(w.local, localIngestDescriptionFlushed)
-				if err == nil {
-					err = w.writer.reopen()
-				}
-			}
-			replyErrCh <- err
-			if err != nil {
-				w.writeErr.Set(err)
-				return
-			}
+	w.totalSize += w.batchSize
+	if len(w.writeBatch) > 0 {
+		if err := w.flushKVs(ctx); err != nil {
+			return errors.Trace(err)
 		}
 	}
 
-	if err = w.writeKVsOrIngest(0); err != nil {
-		w.writeErr.Set(err)
-		return
-	}
 	if w.writer != nil {
-		if err := w.writer.ingestInto(w.local, 0); err != nil {
-			w.writeErr.Set(err)
+		meta, err := w.writer.close()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		w.writer = nil
+		if meta != nil && meta.totalSize > 0 {
+			return w.local.addSST(ctx, meta)
 		}
 	}
+
+	return nil
 }
 
-func (w *Writer) writeKVsOrIngest(desc localIngestDescription) error {
-	if w.writer != nil {
-		if err := w.writer.writeKVs(&w.writeBatch); err != errorUnorderedSSTInsertion {
-			return err
-		}
-	}
+func (w *Writer) Close(ctx context.Context) error {
+	defer w.kvBuffer.destroy()
+	defer w.local.localWriters.Delete(w)
+	err := w.flush(ctx)
+	// FIXME: in theory this line is useless, but In our benchmark with go1.15
+	// this can resolve the memory consistently increasing issue.
+	// maybe this is a bug related to go GC mechanism.
+	w.writeBatch = nil
+	return err
+}
 
-	// if write failed only because of unorderedness, we immediately ingest the memcache.
-	immWriter, err := newSSTWriter(w.genSSTPath())
+func (w *Writer) flushKVs(ctx context.Context) error {
+	writer, err := w.createSSTWriter()
 	if err != nil {
-		return err
+		return errors.Trace(err)
 	}
-	defer immWriter.cleanUp()
-
-	if err = immWriter.writeKVs(&w.writeBatch); err != nil {
-		return err
+	sort.Slice(w.writeBatch[:w.batchCount], func(i, j int) bool {
+		return bytes.Compare(w.writeBatch[i].Key, w.writeBatch[j].Key) < 0
+	})
+	writer.minKey = append(writer.minKey[:0], w.writeBatch[0].Key...)
```

Comment:
seems we can move `writer.minKey` into `writeKVs`, just as `maxKey`

Reply:
No. `appendRowsSorted` will directly call `writeKVs` multi times, but only the first call should set the `minKey`